README
NAME
    Froody - Yet another XML web API framework

SYNOPSIS
      bash$ lwp-request 'http://127.0.0.1:4242/?method=froody.demo.localtime&time=258244200'
      <?xml version="1.0" encoding="utf-8"?>
      <rsp stat="ok">
            <time now="Wed Mar  8 22:30:00 1978">
              <day>8</day>
              <daylight>0</daylight>
              <dow>3</dow>
              <doy>66</doy>
              <hour>22</hour>
              <minute>30</minute>
              <month>2</month>
              <second>0</second>
              <year>78</year>
            </time>
      </rsp>

    See Froody::QuickStart for a better introduction.

DESCRIPTION
    Froody is a framework that can be used to easily create both server and
    clients for making remote API calls across the web.

    Froody communicates by the AJAX friendly calling convention of passing
    parameters to methods via CGI parameters and returning XML data
    structures as a response. The Froody framework handles all the nastyness
    of dealing with the CGI and XML for you, so all you really have to worry
    about in your Perl code is actually getting the job done. In particular,
    it lets you define a strict spec for what will be returned simply by
    example, making new methods quick and easy to write.

  What Documentation Is There?
    First, before you do anything else, you want to read Froody::QuickStart.
    This is the basic tutorial which'll get you up and running and give you
    and idea of what Froody is all about.

    Once you've read that you probably want to read one of the following

    Froody::API::XML
        How to quickly write an XML specification for Froody Methods.

    Froody::Implementation
        What you need to know in order to implement the methods you
        specified reading the above.

    Froody::DataFormats
        A quick review of the data formats used by Froody

    Froody::Server
        Setting up a Froody Server

BUGS
    Please report any bugs you find via the CPAN RT system.
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Froody>

AUTHOR
    Copyright Fotango 2005. All rights reserved.

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    This module has been worked on by the following people:

      Stig Brautaset   <stig@brautaset.org>
      Nicholas Clark   <nclark@fotango.com>
      Mark Fowler      <mark@twoshortplanks.com>
      Chia-liang Kao   <clkao@fotango.com>
      Tom Insam        <tinsam@fotango.com>
      Norman Nunley    <nnunley@fotango.com>

    You can reach the current maintainers by emailing us at
    "cpan@fotango.com", but if you're reporting bugs *please* use the RT
    system mentioned above so we can track the issues you report.

SEE ALSO
    The "examples" directory that ships with this distribution.



Changes
42.041_2

  * Use Class::Accessor::Chained::Fast instead of Class::Accessor in
    Froody::Implementation.

42.041_1

  * Work around broken behaviour in old versions of XML::LibXML.
  * add json callback hook, and test json response.
  * Use JSON::XS, not JSON::Syck - we prefer it because it does Unicode
    correctly among other things.
  * trim BOM from paramters.
  * also decode paramter names from utf-8.
  * catch non-utf8 params better.

42.034
  * Run some tests only if certain optional modules are installed.
  * Better froodydoc (written using JS::Standalone)
  * Fix irritating pod::coverage syntax glitch
  * don't 500 on errors in the post_process handler. Specifically, xml shape errors.
  * Make sure we UTF-8 encode all code paths.
  * Update the bin/froody script to work with modern calling techniques.
  * You can now get at the params _after_ argument chjecking, but before dispatch.
  * All froody servers should support JSON response types.
  * kill deprecated functionality
  * kill untested (and therefore unnused, right?) method
  * Do not load the spec many times. This saves us a _lot_ of time on startup.
  * Fixing the error shape bug.
  * Use findvalue() instead of findnodes/nodeValue gave us an observed 20%
    improvement in loading a particular API module used internally.
  * Reduced copy overhead on walking the reflection structures
  * Refactoring extraction of the XML structure by adding another
    pass, but simplifying the node construction. Also, this fixes a
    regression where last defined node in the example spec wins,
    rather than the first, which was the old behaviour.

42.011
  * Amend arguments to allow argument type aliases
  * Fix the documentation for the Walker class.
  * Strip whitespace in attribute _NAMES_, not attribute values.
  * Can set timeout for SimpleClient when creating the client
  * Fixing the way that the XML API is loaded. We should not be holding on to
    the DOM tree for each example response.
  * Improved XML api code that constructs examples as it creates the specification.
  * fix froody error shapes -- they now pass across the wire for reflection
  * remove XML dom node storage inside of froody
  * avoid cloning data when not necessary

42.010
  * delete undef arguments early instead of trying to encode them
  * handle the case where tempfile is undef during an upload
  * unbreak get_methods() (fixes froodydoc etc.)

42.009
  * fix UTF8 handling for the standalone server and simpleclient
  * remove dependency on JSON
  * (froodydoc binary is currently broken, but the UTF8 fix is important
    enough to release anyway)

42.008
  * added simple (fast) JSON-based client
  * made passing TODO tests not be TODO tests anymore

42.007
  * fixed MANIFEST.

42.006
  * validate terse data structures properly whether they are 
    actually passed over http as XML or simply returned directly
  * add support for returning JSON instead of XML
  * implement a pluggable architecture, useful for sessions management etc
  * added froody.reflection.getSpecification
  * lazy loading of method specs
  * added pluggable (and chainable) argument validators
  * added tests for weird edge cases
  * fixed a few spelling errors in errors

NOTE: Froody is still in a bit of flux, as development is driving
refactorings and new features.


42.005:
  * Allow underscores in method names
  * Allow method names with no dots

42.004:
  * regenerate dist with latest Module::Build::Kwalitee to allow
    building with MakeMaker 

42.003:
  * Fix corelist example so it returns serveradmin stuff
  * added a 'remaining' argument type that holds all non-declared arguments
    passed to the method
  * added examples to MANIFEST (oops)

42.002:
  * Fix broken meta-test (t/t.t)

42.001:
  * Fix to Build.PL

42:
  * Initial release


lib/Froody.pm
package Froody;

use strict;
use warnings;

# okay, you got me.  This doesn't really do anything but hold the universal
# VERSION for Froody so you can say "use Froody 42.1" and it'll go bang if it's
# not installed.  Which it will, 'cos we ain't written it yet
our $VERSION = "42.041_2";

=head1 NAME

Froody - Yet another XML web API framework

=head1 SYNOPSIS

  bash$ lwp-request 'http://127.0.0.1:4242/?method=froody.demo.localtime&time=258244200'
  <?xml version="1.0" encoding="utf-8"?>
  <rsp stat="ok">
	<time now="Wed Mar  8 22:30:00 1978">
	  <day>8</day>
	  <daylight>0</daylight>
	  <dow>3</dow>
	  <doy>66</doy>
	  <hour>22</hour>
	  <minute>30</minute>
	  <month>2</month>
	  <second>0</second>
	  <year>78</year>
	</time>
  </rsp>

See L<Froody::QuickStart> for a better introduction.

=head1 DESCRIPTION

Froody is a framework that can be used to easily create both a server and a
client for making remote API calls across the web.

Froody communicates by the AJAX friendly calling convention of passing
parameters to methods via CGI parameters and returning XML data structures as a
response.  The Froody framework handles all the nastyness of dealing with the
CGI and XML for you, so all you really have to worry about in your Perl code is
actually getting the job done.  In particular, it lets you define a strict spec
for what will be returned simply by example, making new methods quick and easy
to write.

=head2 What Documentation Is There?

First, before you do anything else, you want to read L<Froody::QuickStart>.
This is the basic tutorial which'll get you up and running and give you
and idea of what Froody is all about.

Once you've read that you probably want to read one of the following

=over

=item Froody::API::XML

How to quickly write an XML specification for Froody Methods.

=item Froody::Implementation

What you need to know in order to implement the methods you
specified reading the above.

=item Froody::DataFormats

A quick review of the data formats used by Froody

=item Froody::Server

Setting up a Froody Server

=back

=head1 BUGS

Please report any bugs you find via the CPAN RT system.
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Froody>

=head1 AUTHOR

Copyright Fotango 2005.  All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

This module has been worked on by the following people:

  Stig Brautaset   <stig@brautaset.org>
  Nicholas Clark   <nclark@fotango.com>
  Mark Fowler      <mark@twoshortplanks.com>
  Chia-liang Kao   <clkao@fotango.com>
  Tom Insam        <tinsam@fotango.com>
  Norman Nunley    <nnunley@fotango.com>

You can reach the current maintainers by emailing us at C<cpan@fotango.com>,
but if you're reporting bugs I<please> use the RT system mentioned above so
we can track the issues you report.

=head1 SEE ALSO

The C<examples> directory that ships with this distribution.

=cut

# return true to keep perl happy
'"This must be Thursday," said Arthur to himself, sinking low over his beer, "I never could get the hang of Thursdays."' 
